Refactor server development documentation (#1629)
getsentry/sentry-docs: src/collections/_documentation/development/server/environment.md

---
title: 'Development Environment'
sidebar_order: 2
---

> This guide steps you through configuring a local development environment for the Sentry server on macOS. If you're using another operating system the instructions are still roughly the same, but we don't maintain any official documention.

Author: People can also just look at the Brewfile and figure it out just as safely. IMO it's better to not suggest they can do something thats 10% supported, and wee can update thhis later for Linux upport. The previous version said "go do what the Dockerfile does", and at least one staff member tried that and struggled.

> This guide will step you through configuring a local development environment for the Sentry server on macOS. If you're using another operating system the instructions are still roughly the same, but we don't maintain any official documention.

Reviewer: Nit: present tense, so "This guide steps you through..."

#### A quick note on Bash vs Zsh...

> macOS Catalina will recommend that you use `zsh` instead of `bash`. While the differences are minimal, this guide primarily covers Bash. Verify your shell by running `echo $SHELL`.

Author: Agreed - I actually dont think we really need this callout - was just preserving it. Only a few places it actually applies (e.g. Volta) and if we just say "use the official docs to install those things" we're probably in a good spot.

> macOS Catalina will recommend that you use `zsh` instead of `bash`. While the differences are minimal, this guide primarily covers Bash. Verify your shell by running `echo $SHELL`.

Author: Aside the reason its _not_ a note is because this pagee was filled with those kinds of callout boxes and it's extremely distracting visually. I tried to remove the alert boxes as most things are just instructions. I kid you not the page was like 20% visually alert boxes (and still has quite a few)

> macOS Catalina will recommend that you use `zsh` instead of `bash`. While the differences are minimal, this guide primarily covers Bash. Verify your shell by running `echo $SHELL`.

Reviewer 2: That works for me as long as you remove this section completely or at least remove the ellipsis from the title 😀

> macOS Catalina will recommend that you use `zsh` instead of `bash`. While the differences are minimal, this guide primarily covers bash. You can verify your shell by typing `echo $SHELL`.

Reviewer: backtick `bash` at end of sentence for consistency?

"Verify your shell ..."

Author: bash is both a command (thus `bash`) as well as program. doesn't need to be consistently in backticks (though probably should be Bash vs bash)

## Clone the Repository

> To get started, clone the repo at [https://github.com/getsentry/sentry](https://github.com/getsentry/sentry) - or, your fork if you are not a member of `@getsentry`.

Author: We dont tell staff to fork the repo (in fact, they should not be doing that, can discuss on Slack why), and we dont really need to describe how GitHub works. This is going to be mostly used by staff in the near-term, and I think it's fair to expect people to learn git outside of these instructions. This was the same intention as the rest of my changes: assume people can figure out the basics of computers on their own.

```bash
git clone https://github.com/getsentry/sentry.git
cd sentry

> ```

Author: I figured we'd take a pass at the second round of changes which talk about sentry vs getsentry - I'm fine with explaining that, or even having a second doc that embeds part of this one for staff, At the very least no reason it can't be public, and we can just ensure that both setups mimic each other.

You're going to be working out of the repository for the remainder of the setup.

## System Dependencies

Install [Homebrew](http://brew.sh), and then run `brew bundle` to install the various system packages as listed in sentry's `Brewfile`.

> One thing that requires manual attention is `docker`, which should have just been installed. Open up Spotlight, search for "Docker" and start it. You should soon see the docker icon in your macOS toolbar. Docker will automatically run on system restarts, so this should be the only time you do this.

Author: I dont think its actually a concern? If you install Docker for mac it tries to update itself almost daily, and I don't think we use any cutting edge features?

> One thing that requires manual attention is `docker`, which should have just been installed. Open up Spotlight, search for "Docker" and start it. You should soon see the docker icon in your macOS toolbar. Docker will automatically run on system restarts, so this should be the only time you do this.

Reviewer 2: Works for me. It means less maintenance overhead for the docs.

> You can verify that docker is running by running `docker ps`. If it doesn't error with something like `Error response from daemon: dial unix docker.raw.sock: connect: connection refused`, you're good to continue.

Reviewer: If a customer does see that error, or another that is similar, should they pause, fix, then try again?

Author: i dont know, but this isnt customer facing docs, its for our developers + external contributors

there's a moderate expectation of "you get paid to do this for a living, figure it out" in these kinds of scenarios

> You can verify that docker is running by running `docker ps`. If it doesn't error with something like `Error response from daemon: dial unix docker.raw.sock: connect: connection refused`, you're good to continue.

Reviewer: cap Docker

## Python

> We utilize [pyenv](https://github.com/pyenv/pyenv) to install and manage python versions. It should have already been installed earlier when you ran `brew bundle`.

Reviewer: If it wasn't installed earlier, manually install? Does this suggest that the install went sideways and the customer should check for other badness?

Author: yes but we dont need to spell it out, see above

To install the required versions of Python you'll need to run `pyenv install`. This will take a while, since your computer is actually compiling python!

> After this, if you type `which python`, you should see something like `/usr/bin/python`... this means `python` will resolve to the system's python. You'll need to make some manual changes to your shell initialization files, if you want your shell to see pyenv's python.

Reviewer 3: > If not, then I need to change my shell init files.

Correct. I'm trying to explain why, rather than just tell people to make those changes. Those changes follow right after that sentence.

> After this, if you type `which python`, you should see something like `/usr/bin/python`... this means `python` will resolve to the system's python. You'll need to make some manual changes to your shell initialization files, if you want your shell to see pyenv's python.

Author: fwiw i want to kill all of this from the setup docs and just put a "Having trouble with pyenv? [GO TO PYENV FAQ PAGE]" so we can make this short and sweet

If it worked, running `which python` should result in something like `/Users/you/.pyenv/shims/python`.

### Virtual Environment

You're now ready to create a python virtual environment. Run:

```bash
python -m pip install virtualenv
python -m virtualenv .venv
```

And activate the virtual environment:

```bash
source .venv/bin/activate
```

If everything worked, running `which python` should now result in something like `/Users/you/sentry/.venv/bin/python`.

## JavaScript

We use [volta](https://github.com/volta-cli/volta) to install and manage the version of Node.ks that Sentry requires. To install Volta run:

```bash

> curl https://get.volta.sh | bash

Reviewer 3: Oh, neat.

Volta intercepts this and automatically downloads and installs the node version in sentry's `package.json`.

## direnv

We use [direnv](https://github.com/direnv/direnv) to automate configuration and constraints. It automatically sets some helpful environment variables for you, activates your virtual environment, and does some simple state checking to guide you towards the expected development environment.

> To install it, run `brew install direnv`. You'll want to add this to the end of your `~/.bashrc` file:

Author: Its not optional - that was one of my goals was to remove every "suggestion". It's just "do this stuff, this is how we support development. if you dont do this stuff, assume all risk."

> To install it, run `brew install direnv`. You'll want to add this to the end of your `~/.bashrc` file:

Reviewer 2: Then we may wanna clarify that and use a stronger language. We should also probably add it to the `Brewfile`?

> To install it, run `brew install direnv`. You'll want to add this to the end of your `~/.bashrc` file:

Author: Stronger language in what way? Think i removed everything that suggested it was optional. Agree re: Brewfile.

> To install it, run `brew install direnv`. You'll want to add this to the end of your `~/.bashrc` file:

Reviewer 2: ```suggestion
You need to install it, and add the following snippet to the end of your `~/.bashrc` file:
```

```bash
eval "$(direnv hook bash)"
```

Or in your `~/.zshrc`:

> ```zsh

Author: i just threw this in here since it was eassy and i know we'll be updating this to look at zsh in the future. can back it out if we think its noisy.

really it should just be "do what direnv says when you install it, as it already tells you to do zsh or bash correctly"

> ```zsh

Reviewer 2: Then I'd rather remove it but leaving the final decision up to you.

PATH="" exec /bin/bash -l
```

Any time the `.envrc` configuration changes (including the first time) you will be prompted to run `direnv allow` before any of the configuration will run.

## Bootstrap Services

> The last step is to run `make bootstrap`. This will take a long time, as it bootstraps not only Sentry and it's dependencies, but starts up related services, and runs database migrations.

Reviewer: grammar - "its dependencies"

The `bootstrap` command does a few things you'll want to know about:

- `sentry init` creates the baseline Sentry configuration in `~/.sentry/`.
- `sentry devservices up` spins up required Docker services (such as Postgres and Clickhouse)

> - `sentry upgrade` runs Postgres migrations, and will also prompt you to create a user. You will want to ensure your first user is designated as **superuser**.

Reviewer 2: Idea: keep an up-to-date postgres image/volume so people can just download it instead of running all the migrations from scratch?

Author: maybe useful for someone whos not _really_ developing on Sentry, but not really useful for staff since migrations change frequently

we'd be better off dumping to .sql and just letting them load that, but then we're just duplicating the migration process
